### DISCUSSION TITLE
how to add pdf render to this whiteboard

### DISCUSSION THREAD

**Discussion by tdtanvn** [excalidraw/excalidraw#2795] [Q&A]
hello all,
I am trying to add pdf.js to render pdf content in this whiteboard and I am facing an issue that first load it is ok but when I touch/start daw then pdf content is clear(or something like that).
image: 
![image](https://user-images.githubusercontent.com/2118292/104795893-73175100-57e4-11eb-8675-c825e0d2b4d4.png)
![image](https://user-images.githubusercontent.com/2118292/104795921-a1952c00-57e4-11eb-9c33-4dc6081395af.png)
 and i want to ask what is a good way to add pdf to this whiteboard?
1. create 2 canvas, one for pdf render and one for whiteboard?
2. use once canvas. ( how to add ?) currently, i added in the constructor
![image](https://user-images.githubusercontent.com/2118292/104796054-cd64e180-57e5-11eb-88be-621c8a24bcbe.png)

full code:
[App.zip](https://github.com/excalidraw/excalidraw/files/5823764/App.zip)

Thank you

**Comment by dwelle** [comment]
We re-render the canvas every render-loop (when edit, or sometimes even just when you interact with the canvas as you saw for yourself). 

Rendering on second canvas probably won't work either because if you put it underneath, I believe it won't be visible because we fill our canvas with the background color (you'd need to set it to `transparent`). Putting it above would work, but you'd need to make your PDF has transparent background.

The alternative is to tie the PDF into the renderer itself. Either [add support for image elements yourself](https://github.com/excalidraw/excalidraw/issues/19), or hardcode it into the render step in some other way (check `renderScene.ts`).

**Comment by tdtanvn** [comment]
Thank you for your reply.

I try to add pdf re-render to renderscence.ts then see that works.
and I have another question. can we disable re-render the canvas every render loop ?. because I have a test draw with ~600 elements. then see typing text is lagged. and free draw very hard to write by pen(Wacom)
and if the content of pdf is big maybe it can not re-render fast.

and are you have a plan to support render PDF?
